### PR TITLE
fix Envoy Fails to Add Router

### PR DESCRIPTION
**Pull request by zirain**
fixes: #272 
Signed-off-by: hejianpeng <hejianpeng2@huawei.com>

wait https://github.com/envoyproxy/gateway/pull/273

### CODE REVIEW

File: internal/xds/translator/testdata/in/xds-ir/empty-hostnames.yaml [hunk 1]
```diff
@@ -0,0 +1,10 @@
+name: "http-route"
+http:
+- name: "first-listener"
+  address: "0.0.0.0"
+  port: 10080
+  routes:
+  - name: "first-route"
```
Comment (danehans) [inline internal/xds/translator/testdata/in/xds-ir/empty-hostnames.yaml:7]: Remove whitespace.

**Review by danehans — CHANGES_REQUESTED** [review]
@zirain I am still seeing the same #272 issue with this PR:

## Envoy Gateway Logs
```
2022-08-29T16:43:11.320Z	INFO	cache/logrwrapper.go:24	nodeID "envoy-eg-6f568884ff-6f4gx" requested type.googleapis.com/envoy.config.listener.v3.Listener[] and known map[]. Diff []	{"runner": "xds-server"}
2022-08-29T16:43:11.320Z	INFO	cache/logrwrapper.go:24	respond type.googleapis.com/envoy.config.listener.v3.Listener[] version "" with version "1"	{"runner": "xds-server"}
2022-08-29T16:43:11.320Z	INFO	cache/logrwrapper.go:24	Sending Response on stream 3 to node envoy-eg-6f568884ff-6f4gx	{"runner": "xds-server"}
```

## Envoy Logs
```
[2022-08-29 16:43:10.292][1][warning][config] [source/common/config/grpc_subscription_impl.cc:126] gRPC config for type.googleapis.com/envoy.config.listener.v3.Listener rejected: Error adding/updating listener(s) listener_default-eg-http_8080: Didn't find a registered implementation for 'envoy.filters.http.router' with type URL: ''
```

The [failed CI run](https://github.com/envoyproxy/gateway/runs/8065947972?check_suite_focus=true) supports these findings.

File: internal/xds/translator/translator.go [hunk 2]
```diff
@@ -52,6 +52,10 @@ func Translate(ir *ir.Xds) (*types.ResourceVersionTable, error) {
 			Domains: httpListener.Hostnames,
 		}
 
+		if len(vHost.Domains) == 0 {
```
Comment (arkodg) [inline internal/xds/translator/translator.go:55]: afaik this shouldnt be needed here, the Gateway API translator should have set this to `*` while creating the IR
https://github.com/envoyproxy/gateway/blob/a4773ec06b675cc4a6ee1220d0c8ece09053471c/internal/ir/xds.go#L84

**Comment by danehans** [comment]
FYI @zirain @arkodg I do not see this issue when I create an HTTPRoute that includes a `hostname`. FOr example:
```
apiVersion: gateway.networking.k8s.io/v1alpha2
kind: HTTPRoute
metadata:
  name: httpbin
spec:
  parentRefs:
  - name: eg
  hostnames:
  - "*.example.com"
  rules:
  - backendRefs:
    - name: httpbin
      port: 80
```

File: internal/xds/translator/translator.go (hunk 2 above)
Comment (arkodg) [reply internal/xds/translator/translator.go:55]: looks like `Validate` is not called today, will raise a PR once https://github.com/envoyproxy/gateway/pull/216 is in to call Validate before performing a `Store`

Comment (arkodg) [reply internal/xds/translator/translator.go:55]: raised https://github.com/envoyproxy/gateway/pull/276

**Comment by zirain** [comment]
@danehans I face this issue with the yaml files your provider [here](https://github.com/envoyproxy/gateway/issues/267#issue-1352727798)

I'm OK to close this PR, which is a kind of `defensive programming`.

**Comment by zirain** [comment]
still face this with https://github.com/envoyproxy/gateway/pull/276 https://github.com/envoyproxy/gateway/pull/264 https://github.com/envoyproxy/gateway/pull/269

```
[2022-08-30 08:10:30.712][1][warning][config] [source/common/config/grpc_subscription_impl.cc:126] gRPC config for type.googleapis.com/envoy.config.route.v3.RouteConfiguration rejected: Proto constraint validation failed (RouteConfigurationValidationError.VirtualHosts[0]: embedded message failed validation | caused by VirtualHostValidationError.Domains: value must contain at least 1 item(s)): name: "route_default-eg-http"
virtual_hosts {
  name: "route_default-eg-http"
  routes {
    match {
      prefix: "/"
      headers {
        name: ":authority"
        string_match {
          exact: "*.example.com"
        }
      }
    }
    route {
      cluster: "cluster_default-httpbin-rule-0-match-0-*.example.com"
    }
  }
}
```

**Comment by zirain** [comment]
replaced by https://github.com/envoyproxy/gateway/pull/286

**Pull request closed without merging**